File: BaekJoon/동적 계획법 1/9252_LCS 2.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Integer.max


fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val first = readLine().chunked(1)
    val second = readLine().chunked(1)

    val row = first.size
    val col = second.size

    val dp = Array(row + 1) { Array(col + 1) { 0 } }

    for (i in 1..row) {
        for (j in 1..col) {
            if (first[i - 1] == second[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
            }
        }
    }

    if (dp[row][col] == 0) println(dp[row][col])
    else {
        val stringBuffer = StringBuffer()
        var x = row
        var y = col

        while (x > 0 && x > 0) {
            if (dp[x][y] == dp[x - 1][y]) x--
            else if (dp[x][y] == dp[x][y - 1]) y--
            else {
                stringBuffer.append(second[y - 1])
                x--
                y--
            }
        }

        println(dp[row][col])
        println(stringBuffer.reverse())
    }
}

File: BaekJoon/DP/9252_LCS 2.swift
let s1 = readLine()!.map { String($0) }
let s2 = readLine()!.map { String($0) }
var dp = [[Int]](repeating: [Int](repeating: 0, count: s1.count + 1), count: s2.count + 1)

// LCS 탐색
for i in 1...s2.count {
    for j in 1...s1.count {
        if s2[i - 1] == s1[j - 1] {
            dp[i][j] = dp[i - 1][j - 1] + 1
        } else {
            dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        }
    }
}

// LCS 역추적
var row = s2.count
var col = s1.count
var lcs = ""
var before = dp[row][col]

if before == 0 { print(before) }
else {
    while row > 0, col > 0 {
        if dp[row][col] == dp[row - 1][col] { row -= 1 }
        else if dp[row][col] == dp[row][col - 1] { col -= 1 }
        else {
            lcs += s1[col - 1]
            row -= 1
            col -= 1
        }
    }
    
    print(before)
    print(String(lcs.reversed()))
}